Kafka: Initial base code for policy enforcement and api

manalibhutiyani opened the pull request:
This is the initial piece of code which lays the foundation for supporting Kafka proxy.
This change specifically defines the policy enforcement rules for Kafka and checks
whether only a single rule is configured on a given port.

Fixes #1633
Signed-off by: Manali Bhutiyani manali@covalent.io

Commit:
Kafka: Initial base code for policy enforcement and api
This is the initial piece of code which lays the foundation for supporting Kafka proxy.
This change specifically defines the policy enforcement rules for Kafka and checks
whether only a single rule is configured on a given port.

Fixes #1633
Signed-off by: Manali Bhutiyani manali@covalent.io

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -207,6 +207,10 @@ type L7Rules struct {
 	//
 	// +optional
 	HTTP []PortRuleHTTP `json:"http,omitempty"`
+	// Kafka specific rules.
```
Insert a blank line here.

manalibhutiyani replied:
done

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -207,6 +207,10 @@ type L7Rules struct {
 	//
 	// +optional
 	HTTP []PortRuleHTTP `json:"http,omitempty"`
+	// Kafka specific rules.
```
Kafka specific -> Kafka-specific

manalibhutiyani replied:
done

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -252,3 +256,25 @@ type PortRuleHTTP struct {
+// PortRuleKafka is a list of Kafka protocol constraints. All fields are
+// optional, if all fields are empty or missing, the rule does not have any
+// effect.
+//
```
Remove this line.

manalibhutiyani replied:
done

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -252,3 +256,25 @@ type PortRuleHTTP struct {
+	// If omitted or empty, all methods are allowed.
+	//
+	// +optional
+	ApiKey string `json:"apikey,omitempty"`
```
I would move this field up, to match the structure of Kafka message headers: APIVersion, then APIKey, then Topic.

rlenglet replied:
Capitalize the JSON field name (`apikey`) using camelcase to be consistent with the rest of the API:
`json:"apikey,omitempty"` -> `json:"apiKey,omitempty"`

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -252,3 +256,25 @@ type PortRuleHTTP struct {
+	// Kafka message.
+	Topic string `json:"topic,omitempty"`
+
+	// Method is an extended POSIX regex matched against the method of a
```
Method -> APIKey
method -> type

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -252,3 +256,25 @@ type PortRuleHTTP struct {
+	ApiVersion string `json:"apiVersion,omitempty"`
+
+	// Topic is an POSIX regex matched against the topic of the
+	// Kafka message.
```
Add: Ignored if the matched request message type doesn't contain any topic.

Commit:
Address hound-bot review comments

Fixes #1633
Signed-off by: Manali Bhutiyani manali@covalent.io

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -252,3 +256,25 @@ type PortRuleHTTP struct {
+	// Method is an extended POSIX regex matched against the method of a
+	// request, e.g. "Produce", "Fetch", "CreateTopic", "DeleteTopic", ...
+	//
+	// If omitted or empty, all methods are allowed.
```
methods -> request types

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -252,3 +256,25 @@ type PortRuleHTTP struct {
+
+	// Topic is an POSIX regex matched against the topic of the
+	// Kafka message.
+	Topic string `json:"topic,omitempty"`
```
Add:
```
//
// +optional
```

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -252,3 +256,25 @@ type PortRuleHTTP struct {
+type PortRuleKafka struct {
+	// ApiVersion is an POSIX regex matched against the api version of the
+	// Kafka message
+	ApiVersion string `json:"apiVersion,omitempty"`
```
Add:
```
//
// +optional
```

rlenglet commented on pkg/policy/api/rule_validation.go:
```diff
@@ -93,6 +93,12 @@ func (pr PortRule) Validate() error {
 
+	// Validate L7 rules
+	if pr.Rules != nil {
+		if (pr.Rules.HTTP != nil) && (pr.Rules.Kafka != nil) {
```
Factorize this validation code into a new `(pr *L7Rule) Validate()` method.

rlenglet replied:
In your new `L7Rule.Validate` method, also validate your new Kafka rules in depth. This is not done for HTTP rules, but that's not an excuse for not doing that for Kafka rules. Define a new `(kr *PortRuleKafka) Validate()` method. Iterate on the `Kafka` list and call that validate method, where you should validate that all the fields are valid regexps (calling Go's `regexp.CompilePOSIX` and checking the returned error should be sufficient).

manalibhutiyani replied:
Added the interface, working on the additional in-depth rule checking part.

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -252,3 +256,25 @@ type PortRuleHTTP struct {
+//
+type PortRuleKafka struct {
+	// ApiVersion is an POSIX regex matched against the api version of the
+	// Kafka message
```
It may be worth specifying that the API version is always "0" or a string representing a positive integer.

rlenglet commented on pkg/policy/l4.go:
```diff
@@ -25,7 +25,8 @@ import (
 type AuxRule struct {
-	Expr string `json:"expr"`
+	Expr   string `json:"expr"`
+	L7type string `json:"L7type"`
```
I don't think `type` is the right term here. Either use the same term as in the L4 filter (`L7Parser`) or `L7Proto`.

manalibhutiyani replied:
done

rlenglet commented on pkg/policy/l4.go:
```diff
@@ -43,6 +44,88 @@ type L4Filter struct {
+func FillPortRuleHTTP(httpRule []api.PortRuleHTTP) []AuxRule {
+	l7rules := []AuxRule{}
+	for _, h := range httpRule {
+		r := AuxRule{}
```
Set the L7 protocol / type / parser here:
```
r := AuxRule{L7Parser: "http"}
```

manalibhutiyani replied:
done

rlenglet commented on pkg/policy/l4.go:
```diff
@@ -43,6 +44,88 @@ type L4Filter struct {
+func FillPortRuleKafka(kafkaRule []api.PortRuleKafka) []AuxRule {
+	l7rules := []AuxRule{}
+	for _, h := range kafkaRule {
+		r := AuxRule{}
```
This can be a one-liner:
```
r := AuxRule{L7Proto: "kafka"}
```

manalibhutiyani replied:
done

rlenglet commented on pkg/policy/l4.go:
```diff
@@ -43,6 +44,88 @@ type L4Filter struct {
+	l7rules := []AuxRule{}
+	for _, h := range kafkaRule {
+		r := AuxRule{}
+		r.L7type = "kafka"
```
Define constants for the "http" and "kafka" values.

manalibhutiyani replied:
done

rlenglet commented on pkg/policy/l4.go:
```diff
@@ -43,6 +44,88 @@ type L4Filter struct {
+	for _, h := range kafkaRule {
+		r := AuxRule{}
+		r.L7type = "kafka"
+
```
Add a TODO comment here. The string you produce below is likely not the final format.

manalibhutiyani replied:
done

rlenglet commented on pkg/policy/l4.go:
```diff
@@ -43,6 +44,88 @@ type L4Filter struct {
+		r.L7type = "kafka"
+
+		if h.ApiVersion != "" {
+			r.Expr = "ApiVersionRegexp(\"" + h.ApiVersion + "\")"
```
You didn't rename the usages of ApiVersion and ApiKey when you renamed them in rule.go. I'd recommend using automatic refactoring tools (e.g. Gogland) instead of doing such refactorings by hand.

manalibhutiyani replied:
done

rlenglet commented on pkg/policy/l4.go:
```diff
@@ -59,52 +142,18 @@ func CreateL4Filter(rule api.PortRule, port api.PortProtocol, direction string,
-					r.Expr += " && "
-				}
-				r.Expr += "MethodRegexp(\"" + h.Method + "\")"
+		if rule.Rules.HTTP != nil {
```
Such a structure:
```
if <condition 1> {
...
} else <condition 2> {
...
}
```
Can be rewritten a bit more elegantly into:
```
switch {
case <condition 1>:
...
case <condition 2>:
...
}
```

manalibhutiyani replied:
done

rlenglet commented on pkg/proxy/proxy.go:
```diff
@@ -488,10 +489,17 @@ func (p *Proxy) CreateOrUpdateRedirect(l4 *policy.L4Filter, id string, source Pr
 	}
 
-	if strings.ToLower(l4.L7Parser) != "http" {
+	if !(strings.ToLower(l4.L7Parser) == "http" ||
```
You should never have to convert to lowercase, and this can be made a little cleaner:
```
switch l4.L7Parser {
case "http":
case "kafka":
    log.Debug("MK in CreateOrUpdateRedirect l4.L7Parser:..returning", l4.L7Parser)
    return nil, fmt.Errorf("unsupported L7 protocol proxy:\"%s\"", l4.L7Parser)
default:
    return nil, fmt.Errorf("unknown L7 protocol \"%s\"", l4.L7Parser)
}

manalibhutiyani replied:
done

Commit:
Address romain's review comments

Fixes #1633
Signed-off by: Manali Bhutiyani manali@covalent.io

Commit:
Hound comment review

Fixes #1633
Signed-off by: Manali Bhutiyani manali@covalent.io

Commit:
A different Validate interface for l7 rule validation

Fixes #1633
Signed-off by: Manali Bhutiyani manali@covalent.io

Commit:
Issue #1633 Fix the rule_test

Fixes #1633
Signed-off by: Manali Bhutiyani manali@covalent.io

Commit:
Add rule validation for kafka. Allow simple string matching and whitelisting model for now

Fixes #1633
Signed-off by: Manali Bhutiyani manali@covalent.io

Commit:
Address nits
Fixes #1633
Signed-off by: Manali Bhutiyani manali@covalent.io

Commit:
Nits. comment changes

Fixes #1633
Signed-off by: Manali Bhutiyani manali@covalent.io

tgraf commented on pkg/policy/api/rule.go:
```diff
@@ -252,3 +258,50 @@ type PortRuleHTTP struct {
+
+// KafkaMaxTopicLen is the maximum character len of a topic.
+const (
+	KafkaMaxTopicLen = 249
```
I think older Kafka versions had longer topic lengths of 255 so let's limit against that and add a small comment here to document this.

manalibhutiyani replied:
done. Also added the comment.

tgraf commented on pkg/policy/api/rule_validation.go:
```diff
@@ -82,6 +82,62 @@ func (e EgressRule) Validate() error {
+func (kr PortRuleKafka) Validate() error {
+	if len(kr.APIKey) > 0 {
+		switch kr.APIKey {
+		case KafkaProduceReq:
```
Are these topic names always guaranteed to be capitalized as defined, e.g. "Produce"?

rlenglet replied:
I don’t understand why you allow only a tiny subset of the API keys. Why do you prevent users from matching on cluster messages such as the *Group messages? You easily can and should support all of Kafka’s message types. I would:
1) define a map of valid API keys, with the strings as lowercase keys and the numeric API keys as values, and make sure that all of the API keys in the Kafka spec are in there;
2) here, just convert the input key into lower case, and lookup in the map;
3) if the key is not found, return an error.

manalibhutiyani replied:
@tgraf @rlenglet I was trying to make this as a simple first step and hence took some basic assumptions. Apologies. Let me make the changes and put them out for review.

manalibhutiyani replied:
Added a map KafkaAPIKeyMap with all keys currently supported by Kafka.

tgraf commented on pkg/policy/api/rule.go:
```diff
@@ -207,6 +208,11 @@ type L7Rules struct {
 	//
 	// +optional
 	HTTP []PortRuleHTTP `json:"http,omitempty"`
+	// Kafka-specific rules.
```
Missing empty line and `//`

manalibhutiyani replied:
done

tgraf commented on pkg/policy/api/rule_validation.go:
```diff
@@ -82,6 +82,62 @@ func (e EgressRule) Validate() error {
+	if len(kr.APIVersion) > 0 {
+		switch kr.APIVersion {
+		case "0":
+		case "1":
```
I think it's the right step to define this as string as it allows us to define ranges of versions later on. Instead of hardcoding these versions I would just pass it through:
```u, err := strconv.ParseUint("42", 10, 16)```
and error out if `ParseUint()` fails

manalibhutiyani replied:
sure

manalibhutiyani replied:
done

tgraf commented on pkg/policy/api/rule_validation.go:
```diff
@@ -82,6 +82,62 @@ func (e EgressRule) Validate() error {
+			return fmt.Errorf("Kafka Topic exceeds maximum len of ",
+				KafkaMaxTopicLen)
+		}
+		// This check allows suffix and prefix matching
```
👍

tgraf commented on pkg/policy/api/rule_validation.go:
```diff
@@ -82,6 +82,62 @@ func (e EgressRule) Validate() error {
+// Validate validates L7 rules
+func (pr *L7Rules) Validate() error {
+	if (pr.HTTP != nil) && (pr.Kafka != nil) {
+		return fmt.Errorf("multiple rules for the same port")
```
Let's do a more specific error to make it crystal clear what the error is:

     "multiple L7 protocol rule types specified in single rule"

manalibhutiyani replied:
done

tgraf commented on pkg/policy/api/rule_validation.go:
```diff
@@ -82,6 +82,62 @@ func (e EgressRule) Validate() error {
+		case KafkaCreateTopicsReq:
+		case KafkaDeleteTopicsReq:
+		default:
+			return fmt.Errorf("Invalid Kafka APIKey :", kr.APIKey)
```
All errors returned should start with lower capital letters so it can be embedded.

rlenglet replied:
If you’r using Errorf, use a proper format string:
`fmt.Errorf(“...: %q”, kr.APIKey)`

manalibhutiyani replied:
done

tgraf commented on pkg/policy/l4.go:
```diff
@@ -25,9 +25,16 @@ import (
 }
 
+// HTTPStr for proxy type http
+const HTTPStr = "http"
```
Let's make it clear that this is a parser type:
```
// L7ParserType is the type used to indicate what L7 parser to use and defines all supported types of L7 parsers
type L7ParserType string

const (
    // ParserTypeHTTP specifies a HTTP parser type
    ParserTypeHTTP L7ParserType = "http"

    [...]
)
```

manalibhutiyani replied:
will do.. but looks like it needs typecasting to string everywhere
l4.L7Parser = string(ParserTypeKafka)

manalibhutiyani replied:
done

tgraf commented on pkg/proxy/proxy.go:
```diff
@@ -488,7 +488,13 @@ func (p *Proxy) CreateOrUpdateRedirect(l4 *policy.L4Filter, id string, source Pr
 	}
 
-	if strings.ToLower(l4.L7Parser) != "http" {
+	switch l4.L7Parser {
```
👍

tgraf commented on pkg/policy/l4.go:
```diff
@@ -59,53 +149,20 @@ func CreateL4Filter(rule api.PortRule, port api.PortProtocol, direction string,
-				r.Expr += "HostRegexp(\"" + h.Host + "\")"
+		switch {
+		case rule.Rules.HTTP != nil:
+			l7rules := FillPortRuleHTTP(rule.Rules.HTTP)
```
You can consider having `FIllPortRuleHTTP` and `FillPortRuleKafka` return both the rules and the parser type so you can write:
```
l4.L7Rules, l4.L7Parser = FillPortRuleHTTP(rule.Rules.HTTP)
```

manalibhutiyani replied:
sure

manalibhutiyani replied:
done

tgraf commented on pkg/policy/l4.go:
```diff
@@ -43,6 +50,89 @@ type L4Filter struct {
+// as a regex into a single string
+// TODO The string formed is likely not the final format, and will
+// evolve as more kafka keys are supported.
+func FillPortRuleKafka(kafkaRule []api.PortRuleKafka) []AuxRule {
```
I think whatever parser and enforcer we have, can directly understand `api.PortRuleKafka` so we can probably skip this step and simply pass in the validated rules.

manalibhutiyani replied:
If you mean the Vulcand proxy router? then not really, it is specific to http and all the router checks assume http requests only  and does not understand Kafka at all. I think we need to write the router from scratch for Kafka, as well as the parser, which is what I am working on.


I am trying to see if I can reuse some Vulcand proxy router functionality, and will leave this one as is for now.

I can change it later if I feel its not really needed.

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -252,3 +258,50 @@ type PortRuleHTTP struct {
+	APIVersion string `json:"apiVersion,omitempty"`
+
+	// APIKey is a string matched against the key of a
+	// request, e.g. "Produce", "Fetch", "CreateTopic", "DeleteTopic", ...
```
Echoing @tgraf’s comment, I suppose the value is case-insensitive? If so, specific that explicitly in this comment.

manalibhutiyani replied:
@rlenglet I was trying to make this as a simple first step and hence took some assumptions. Let me make the changes and put them out for review.

manalibhutiyani replied:
done

rlenglet commented on pkg/policy/api/rule.go:
```diff
@@ -252,3 +258,50 @@ type PortRuleHTTP struct {
+	Topic string `json:"topic,omitempty"`
+}
+
+// KafkaProduceReq, KafkaFetchReq, KafkaCreateTopicsReq, KafkaDeleteTopicsReq
```
Why would you limit to a specific subset of API keys? I don’t see any reason why you couldn’t accept all API keys defined in the spec.

manalibhutiyani replied:
Changed. We now allow all keys as per spec, in a case-insensitive manner
https://kafka.apache.org/protocol#protocol_api_keys

rlenglet commented on pkg/policy/api/rule_validation.go:
```diff
@@ -82,6 +82,62 @@ func (e EgressRule) Validate() error {
+
+	if len(kr.Topic) > 0 {
+		if len(kr.Topic) > KafkaMaxTopicLen {
+			return fmt.Errorf("Kafka Topic exceeds maximum len of ",
```
Same here, use a proper format string.

manalibhutiyani replied:
done .

Commit:
Issue #1633 Make rule validation more solid by checking all
APIkeys using a map and other review comments

Fixes #1633
Signed-off by: Manali Bhutiyani manali@covalent.io

Commit:
Issue #1633 make the comments more clear
Fixes #1633
Signed-off by: Manali Bhutiyani manali@covalent.io

Commit:
Issue #1633 Make the api rule comments more pretty

Fixes #1633
Signed-off by: Manali Bhutiyani manali@covalent.io

Commit:
Issue #1633 : HTTPStr --> ParserTypeHTTP in rule_test

Fixes #1633
Signed-off by: Manali Bhutiyani manali@covalent.io

Commit:
Fix rule_test validation

Fixes #1633
Signed-off by: Manali Bhutiyani manali@covalent.io

rlenglet commented on pkg/policy/api/rule_validation.go:
```diff
@@ -82,6 +82,55 @@ func (e EgressRule) Validate() error {
+	}
+
+	if len(kr.APIVersion) > 0 {
+		_, err := strconv.ParseUint(kr.APIVersion, 10, 16)
```
Just a minor nitpick:
```
if _, err := ...; err != nil {
...
}
```

rlenglet commented on pkg/policy/api/rule_validation.go:
```diff
@@ -82,6 +82,55 @@ func (e EgressRule) Validate() error {
+		}
+		// This check allows suffix and prefix matching
+		// for topic.
+		if KafkaTopicValidChar.MatchString(kr.Topic) == false {
```
Just two minor nitpicks:
```
if !KafkaTopicValidChar.MatchString(kr.Topic) {
    return errors.New("invalid Kafka topic")
}
```

rlenglet commented on pkg/policy/api/rule_validation.go:
```diff
@@ -82,6 +82,55 @@ func (e EgressRule) Validate() error {
+// Validate validates L7 rules
+func (pr *L7Rules) Validate() error {
+	if (pr.HTTP != nil) && (pr.Kafka != nil) {
+		return fmt.Errorf("multiple L7 protocol rule types specified in single rule")
```
Minor nitpick:
in single rule -> for the same L4 port (???)

rlenglet commented on pkg/proxy/proxy.go:
```diff
@@ -488,7 +488,13 @@ func (p *Proxy) CreateOrUpdateRedirect(l4 *policy.L4Filter, id string, source Pr
+		log.Debug("Kafka Parser not supported yet")
+		return nil, fmt.Errorf("unsupported L7 protocol proxy:\"%s\"", l4.L7Parser)
+	default:
 		return nil, fmt.Errorf("unknown L7 protocol \"%s\"", l4.L7Parser)
```
Minor nitpick: `\"%s\"` should be replaced with `%q`.

rlenglet commented on pkg/proxy/proxy.go:
```diff
@@ -488,7 +488,13 @@ func (p *Proxy) CreateOrUpdateRedirect(l4 *policy.L4Filter, id string, source Pr
+	case policy.ParserTypeKafka:
+		// TODO We need to remove this once Kakfa parser and router support is in.
+		log.Debug("Kafka Parser not supported yet")
+		return nil, fmt.Errorf("unsupported L7 protocol proxy:\"%s\"", l4.L7Parser)
```
2 minor nitpicks:
`\"%s\"` should be replaced with `%q`
Add a space after the semi-colon `:`.

rlenglet reviewed (approved):
LGTM! This is great. I only have a few nitpicks. You may choose to address them in a subsequent PR. They are not blocking.
